### PR TITLE
Vectorize block_diag to avoid inner loops over block elements

### PR DESCRIPTION
**Pull request by kburns**
This PR modifies block_diag to avoid inner loops over the block entries via vectorization and early conversions to COO.  This results in comparable performance with small dense blocks, but better performance with large dense blocks or sparse blocks.

Before (commit 90cfcbd1291930929736b7c8d182168a456bd8d6):

```
[100.00%] ··· sparse.BlockDiagDenseConstruction.time_block_diag                                       ok
[100.00%] ··· ============== ============
               num_matrices              
              -------------- ------------
                   1000       8.01±0.3ms 
                   5000       37.3±0.9ms 
                  10000        79.3±1ms  
                  15000        122±4ms   
                  20000        164±2ms   
              ============== ============

[100.00%] ··· sparse.BlockDiagSparseConstruction.time_block_diag                                      ok
[100.00%] ··· ============== ==========
               num_matrices            
              -------------- ----------
                   100        13.5±1ms 
                   500        69.5±1ms 
                   1000       150±3ms  
                   1500       221±7ms  
                   2000       296±4ms  
              ============== ==========
```

After (commit bfa3269d108be255aab7420f0be20ee5662155a1):

```
[100.00%] ··· sparse.BlockDiagDenseConstruction.time_block_diag                                       ok
[100.00%] ··· ============== ============
               num_matrices              
              -------------- ------------
                   1000       7.57±0.1ms 
                   5000        38.9±1ms  
                  10000       75.1±0.9ms 
                  15000        115±2ms   
                  20000        158±8ms   
              ============== ============

[100.00%] ··· sparse.BlockDiagSparseConstruction.time_block_diag                                      ok
[100.00%] ··· ============== ============
               num_matrices              
              -------------- ------------
                   100         995±30μs  
                   500        4.98±0.1ms 
                   1000       10.2±0.4ms 
                   1500       15.6±0.9ms 
                   2000       21.2±0.6ms 
              ============== ============
```

### CODE REVIEW

File: scipy/sparse/construct.py [hunk 1]
```diff
@@ -672,18 +672,25 @@ def block_diag(mats, format=None, dtype=None):
...
+        else:
+            a_row, a_col = np.divmod(np.arange(nrows*ncols), ncols)
+            row.append(a_row + r_idx)
+            col.append(a_col + c_idx)
+            data.append(a.ravel())
         r_idx = r_idx + nrows
         c_idx = c_idx + ncols
+    row = np.concatenate(row)
+    col = np.concatenate(col)
+    data = np.concatenate(data)
```
Comment (jjerphan) [inline scipy/sparse/construct.py:693]: Why don't you use `np.array` directly?

**Review by jjerphan — COMMENTED** [review]
Looks like those are nice results @kburns! Thank you for your contribution!
I'll merge after the end of this discussion.

File: scipy/sparse/construct.py (hunk 1 above)
Comment (kburns) [reply scipy/sparse/construct.py:693]: Here `row`, `col`, and `data` are lists that are accumulating (potentially different length) arrays from each block, and then concatenating them all together at the end.  You could preallocate arrays for these, but that would require an additional loop over the blocks to determine the total number of nonzeros and logic to determine the output datatype from all of the block datatypes.  I'm not sure about the performance difference, but this way that's all handled by `concatenate`.

Comment (jjerphan) [reply scipy/sparse/construct.py:693]: OK, thank you, it makes sense.

**Pull request merged**